flake.nix

```diff
@@ -0,0 +1,182 @@
+{
+  description = "Compiling to categories";
+
+  nixConfig = {
+    ## https://github.com/NixOS/rfcs/blob/master/rfcs/0045-deprecate-url-syntax.md
+    extra-experimental-features = ["no-url-literals"];
+    extra-trusted-public-keys = [
+      "cache.garnix.io:CTFPyKSLcx5RMJKfLo5EEPUObbA78b0YQ2DTCJXqr9g="
+    ];
+    extra-trusted-substituters = ["https://cache.garnix.io"];
+    ## Isolate the build.
+    registries = false;
+    sandbox = true;
+  };
+
+  outputs = inputs: let
+    pname = "concat";
+
+    defaultCompiler = "ghc902";
+    supportedGhcVersions = [
+      "ghc8107"
+      "ghc902"
+      "ghc924"
+      # These don’t work yet.
+      # "ghc942" "ghcHEAD"
+    ];
+    excludedPackages = [];
+    noCheckPackages = [
+      # Only works on Linux and Darwin, and I’m not sure how to expose the
+      # `open` executable on Nix/darwin.
+      "concat-graphics"
+      # Just fails.
+      "concat-hardware"
+      # TODO: Tests pass with `cabal new-test`, but not with `cabal test`, and
+      #       I’m guessing the latter is what `callCabal2nix` uses.
+      "concat-plugin"
```

Comment:
I would really like to remove this line – i.e., get the gold-tests working via Nix.

I’m ok with merging before that happens, but this prevents us from replacing GH Actions with Garnix.

Reply:
@sellout What exactly fails with the gold tests?

Reply:
One thing I can think of is that `misc-examples` and `misc-trace` aren't proper test suites.

Just turned them into executables in a0f2aff0a3a9f6cd84e45ae9faa9e5525e9ecb51

Reply:
Ah, great – the same should be done for `graphics-trace` in `concat-graphics`.

Then we should also expose the executables in the flake 😄

Reply:
Your executable change fixes some of what I was running into, but there are still a few issues. Here’s what happens on GHC 9.0.2 (the default):
```
$ nix develop github:con-kitty/concat/updated-flake#ghc902_concat-plugin
⋮
❄️$ cabal test gold-tests
⋮
[3 of 4] Compiling BasicTests       ( test/BasicTests.hs, dist/build/gold-tests/gold-tests-tmp/BasicTests.o )
[4 of 4] Compiling Main             ( test/GoldTests.hs, dist/build/gold-tests/gold-tests-tmp/Main.o )
Linking dist/build/gold-tests/gold-tests ...
/nix/store/…-clang-wrapper-11.1.0/bin/ld: line 256: 51086 Segmentation fault: 11  /nix/store/…-cctools-binutils-darwin-973.0.1/bin/ld ${extraBefore+"${extraBefore[@]}"} ${params+"${params[@]}"} ${extraAfter+"${extraAfter[@]}"}
clang-11: error: linker command failed with exit code 139 (use -v to see invocation)
`cc' failed in phase `Linker'. (Exit code: 139)
❄️$ cabal v2-test gold-tests
⋮
5 out of 83 tests failed (0.14s)

Test suite gold-tests: FAIL
Test suite logged to:
/…/build/aarch64-osx/ghc-9.0.2/concat-plugin-0.3.0.0/t/gold-tests/test/concat-plugin-0.3.0.0-gold-tests.log
0 of 1 test suites (0 of 1 test cases) passed.
Error: cabal: Tests failed for test:gold-tests from concat-plugin-0.3.0.0.

❄️$ 
```
`cabal test` falls during linking, but `cabal v2-test` builds (then fails some test cases). Once the linking step succeeds, `cabal test gold-tests` will thereafter have the same result as `cabal v2-test gold-tests`.

When replicating this issue, I noticed a “bug”. Everything gets built with the correct GHC, but if you enter a dev shell (as above) the GHC that’s exposed in the shell is _always_ the default (currently GHC 9.0.2). So,
 ```
$ nix develop github:con-kitty/concat/updated-flake#ghc8107_concat-plugin
⋮
❄️$ cabal test gold-tests
```
will build and run tests with GHC 9.0.2, not 8.10.7. I tried a few things locally to fix that, but didn’t get anything that works.

However `nix build github:con-kitty/concat/updated-flake#ghc8107_concat-plugin` builds just fine, no linking error, unlike 9.0.2 – and the tests pass on 8.10.7.

Reply:
Ok, I pushed a commit that changed the other things that should be `executable` and then enables all checks (which will still result in some builds failing).

However, I don’t think any of the failing builds are depended on by Categorifier 😁

Reply:
So, to summarize [the current state](https://github.com/compiling-to-categories/concat/pull/111/checks?check_run_id=11691926880):
- `concat-hardware` panics during tests (on all GHCs)
- `concat-plugin` fails tests _only_ on GHC 9.0.2
- all of the other failures are _probably_ fine, but are blocked by the above failures.

Reply:
I just enabled darwin builds, too, so we’ll see how that goes.

Reply:
The "5 out of 83 tests failed" are probably false negatives: You can see that the github ci on master also fails similarly, as the output is equivalent but similar. I'm guessing a slight skew in base versions, leading to different inlining results. If that's the case here (of which I'm pretty sure), we can just re-generate the gold output and be done with it.